src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc

```diff
@@ -418,46 +606,47 @@ void PickFirst::PickFirstSubchannelData::ProcessConnectivityChangeLocked(
   //    select in place of the current one.
   // If the subchannel is READY, use it.
   if (new_state == GRPC_CHANNEL_READY) {
-    subchannel_list()->set_in_transient_failure(false);
+    subchannel_list_->in_transient_failure_ = false;
     ProcessUnselectedReadyLocked();
     return;
   }
   // If we haven't yet seen the initial connectivity state notification
   // for all subchannels, do nothing.
-  if (!subchannel_list()->AllSubchannelsSeenInitialState()) return;
+  if (!subchannel_list_->AllSubchannelsSeenInitialState()) return;
   // If we're still here and this is the initial connectivity state
   // notification for this subchannel, that means it was the last one to
   // see its initial notification.  Start trying to connect, starting
   // with the first subchannel.
   if (!old_state.has_value()) {
-    subchannel_list()->subchannel(0)->ReactToConnectivityStateLocked();
+    subchannel_list_->subchannels_.front().ReactToConnectivityStateLocked();
     return;
   }
   // Ignore any other updates for subchannels we're not currently trying to
   // connect to.
-  if (Index() != subchannel_list()->attempting_index()) return;
+  if (Index() != subchannel_list_->attempting_index_) return;
   // React to the connectivity state.
   ReactToConnectivityStateLocked();
 }
 
-void PickFirst::PickFirstSubchannelData::ReactToConnectivityStateLocked() {
-  PickFirst* p = static_cast<PickFirst*>(subchannel_list()->policy());
+void PickFirst::SubchannelList::SubchannelData::
+    ReactToConnectivityStateLocked() {
+  PickFirst* p = subchannel_list_->policy_.get();
   // Otherwise, process connectivity state.
-  switch (connectivity_state().value()) {
+  switch (connectivity_state_.value()) {
     case GRPC_CHANNEL_READY:
       // Already handled this case above, so this should not happen.
       GPR_UNREACHABLE_CODE(break);
     case GRPC_CHANNEL_TRANSIENT_FAILURE: {
       // Find the next subchannel not in state TRANSIENT_FAILURE.
       // We skip subchannels in state TRANSIENT_FAILURE to avoid a
       // large recursion that could overflow the stack.
-      PickFirstSubchannelData* found_subchannel = nullptr;
+      SubchannelData* found_subchannel = nullptr;
       for (size_t next_index = Index() + 1;
-           next_index < subchannel_list()->num_subchannels(); ++next_index) {
-        PickFirstSubchannelData* sc = subchannel_list()->subchannel(next_index);
-        GPR_ASSERT(sc->connectivity_state().has_value());
-        if (sc->connectivity_state() != GRPC_CHANNEL_TRANSIENT_FAILURE) {
-          subchannel_list()->set_attempting_index(next_index);
+           next_index < subchannel_list_->size(); ++next_index) {
```

Comment:
645-667 seem to use more of `subchannel_list_` - so maybe there should be a "find_next_subchannel_not_transient_failure` method in subchannel list class?

Reply:
This isn't a method that would ever need to be called from anywhere else, so I don't see any real benefit in moving it to a separate function.